clang/test/CodeGenSYCL/kernel-early-optimization-pipeline.cpp

```diff
@@ -22,4 +23,17 @@
 //
 // RUN: %clang_cc1 -O2 -fsycl-is-device -triple spir64-unknown-unknown %s -mdebug-pass Structure -emit-llvm -fno-sycl-early-optimizations -o /dev/null 2>&1 | FileCheck %s --check-prefix=CHECK-NOEARLYOPT
 // CHECK-NOEARLYOPT-NOT: ConstantMergePass1
+// CHECK-NOEARLYOPT-NOT: SYCLOptimizeBarriersPass
 // CHECK-NOEARLYOPT: SYCLMutatePrintfAddrspacePass
+
+// RUN: %clang_cc1 -O0 -fsycl-is-device -triple spir64-unknown-unknown %s -mdebug-pass Structure -emit-llvm -o /dev/null 2>&1 | FileCheck %s --check-prefix=CHECK-O0
+// CHECK-O0-NOT: SYCLOptimizeBarriersPass
+
+template <typename name, typename Func>
+void kernel(const Func &f) __attribute__((sycl_kernel)) {
+  f();
+}
+
+void bar() {
+  kernel<class MyKernel>([=]() {});
+}
```

Comment:
I find it really weird that the pass manager doesn't report this pass if the source is empty, however I'm not an expert in this. Can we simplify the code and explain why it is required, it is dummy anyway
```suggestion
// Some passes don't run on empty code
__attribute__((sycl_device)) void bar() {
}
```
If a kernel is required please use https://github.com/intel/llvm/blob/sycl/sycl/doc/developer/ContributeToDPCPP.md#dpc-clang-fe-tests .

Reply:
Applied. So passes registered via `registerOptimizerLastEPCallback` are not being run fir the empty test. Passes registered like this are being run at the end of a normal pipeline via `invokeOptimizerLastEPCallbacks`  and ... I haven't yet finished with the debugging, but it seems like it's module-size sensitive.